### PR TITLE
[TACHYON-108] Remote write support

### PR DESCRIPTION
**Pull request by gpang**
https://tachyon.atlassian.net/browse/TACHYON-108

This PR adds support for remote writes in Tachyon. The implementation is very simple, which means it NOT optimized at all. There are many opportunities to make everything faster and more efficient. Optimizations can be implemented in future PRs.

(LocalBlockOutStream.java was pretty much copied from BlockOutStream.java, but github doesn't show that)

### CODE REVIEW

**Comment by calvinjia** [comment]
Jenkins, retest this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

File: common/src/main/java/tachyon/network/protocol/RPCBlockWriteRequest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ */
+public class RPCBlockWriteRequest extends RPCRequest {
+  private final long mUserId;
+  private final long mBlockId;
+  private final long mOffset;
+  private final long mLength;
+  private final DataBuffer mData;
+
+  public RPCBlockWriteRequest(long userId, long blockId, long offset, long length,
+                              DataBuffer data) {
```
Comment (haoyuan) [inline common/src/main/java/tachyon/network/protocol/RPCBlockWriteRequest.java]: format

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java [hunk 2]
```diff
@@ -16,238 +16,62 @@
...
-        mBuffer.clear();
-      }
-      appendCurrentBuffer(b, off, len);
-    } else if (len > 0) {
-      // Write the data to the buffer, and not directly to the mapped file.
-      mBuffer.put(b, off, len);
-    }
-
-    mWrittenBytes += len;
+  public BlockOutStream(TachyonFile tachyonFile, WriteType opType, TachyonConf tachyonConf) {
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: is `public` or `private`, given we have factory method?

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * <code>LocalBlockOutStream</code> implementation of TachyonFile. This class is not client facing.
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: NIT: I think the javadoc here is both confusing and probably wrong. Better improve the javadoc here.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    mBlockIndex = blockIndex;
+    mBlockCapacityByte = mFile.getBlockSizeByte();
+    mBlockId = mFile.getBlockId(mBlockIndex);
+    mBlockOffset = mBlockCapacityByte * blockIndex;
+
+    mCanWrite = true;
+
+    if (!mTachyonFS.hasLocalWorker()) {
+      mCanWrite = false;
+      String msg = "The machine does not have any local worker.";
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: (1) why `msg` is needed?
(2) Assigning `mCanWrite` seems useless.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockWriter.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import java.net.InetSocketAddress;
+
+import com.google.common.base.Throwables;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * The interface to read remote block from data server.
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockWriter.java]: read -> write

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java [hunk 6]
```diff
@@ -16,238 +16,62 @@
...
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import tachyon.Constants;
 import tachyon.conf.TachyonConf;
-import tachyon.util.CommonUtils;
 
 /**
  * <code>BlockOutStream</code> implementation of TachyonFile. This class is not client facing.
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: No more an implementation since this is an abstract class

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+      throw new IOException(msg);
+    }
+    mLocalFilePath = mTachyonFS.getLocalBlockTemporaryPath(mBlockId, initialBytes);
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Better leave a TODO and state that LocalFileBlockWriter will be used instead of using java file API directly?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+import com.google.common.primitives.Ints;
+
+import tachyon.Constants;
+import tachyon.conf.TachyonConf;
+import tachyon.util.CommonUtils;
+
+/**
+ * <code>RemoteBlockOutStream</code> implementation of TachyonFile. This class is not client facing.
+ */
+public class RemoteBlockOutStream extends BlockOutStream {
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: `LocalBlockOutStream` is package-private (I thought that is by purpose, isn't it?) 
why this `RemoteBlockOutStream` is public?

File: clients/unshaded/src/main/java/tachyon/client/TachyonFile.java [hunk 9]
```diff
@@ -497,7 +497,7 @@ boolean recache(int blockIndex) throws IOException {
       int bufferBytes =
           (int) mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB) * 4;
       byte[] buffer = new byte[bufferBytes];
-      bos = new BlockOutStream(this, WriteType.TRY_CACHE, blockIndex, mTachyonConf);
+      bos = new LocalBlockOutStream(this, WriteType.TRY_CACHE, blockIndex, mTachyonConf);
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFile.java]: should this be `LocalBlockOutStream` always? If `!mTachyonFS.hasLocalWorker()`, `new LocalBlockOutStream` will throw exceptions

File: common/src/main/java/tachyon/network/protocol/RPCMessage.java [hunk 10]
```diff
@@ -31,7 +31,10 @@
 
   // The possible types of RPC messages.
   public enum Type implements EncodedMessage {
-    RPC_BLOCK_REQUEST(0), RPC_BLOCK_RESPONSE(1);
+    RPC_BLOCK_REQUEST(0),
```
Comment (apc999) [inline common/src/main/java/tachyon/network/protocol/RPCMessage.java]: NIT: should this better be `RPC_BLOCK_READ_REQUEST` and `RPC_BLOCK_RESPONSE`? Maybe in another PR

**Comment by apc999** [comment]
Put a bunch of comments. Thanks for the patch!

**Comment by gpang** [comment]
Jenkins, test this please.

File: common/src/main/java/tachyon/network/protocol/RPCBlockWriteRequest.java (hunk 1 above)
Comment (gpang) [inline common/src/main/java/tachyon/network/protocol/RPCBlockWriteRequest.java]: done.

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 2 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: I made it protected, so subclasses can call it.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 3 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: I updated the comments.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 4 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: I removed msg, and don't assign mCanWrite.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockWriter.java (hunk 5 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockWriter.java]: done.

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 6 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Updated comments.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 7 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Added the TODO.

File: clients/unshaded/src/main/java/tachyon/client/TachyonFile.java (hunk 9 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFile.java]: Yes, this can be either a Local one or Remote one. I changed it so it will choose accordingly. I had originally thought this was only possible locally, but remote should be possible. Thanks!

File: common/src/main/java/tachyon/network/protocol/RPCMessage.java (hunk 10 above)
Comment (gpang) [inline common/src/main/java/tachyon/network/protocol/RPCMessage.java]: They can be renamed in a different PR. I left a few TODO's to rename them.

**Comment by gpang** [comment]
Thanks @haoyuan and @apc999 for the reviews! I addressed your comments.

This PR can be continued to be reviewed, but it will be updated again after the big worker WIP is merged into master, to use the new APIs. This PR should only be merged in after that.

**Comment by gpang** [comment]
This PR is now rebased onto master.

File: clients/unshaded/src/main/java/tachyon/client/TachyonFS.java [hunk 11]
```diff
@@ -642,23 +642,7 @@ public synchronized ClientFileInfo getFileStatus(int fileId, boolean useCachedMe
...
-    if (!localTempFolder.exists()) {
-      if (localTempFolder.mkdirs()) {
-        CommonUtils.changeLocalFileToFullPermission(localTempFolder.getAbsolutePath());
-        LOG.info("Folder {} was created!", localTempFolder);
-      } else {
-        throw new IOException("Failed to create folder " + localTempFolder);
-      }
-    }
-
+    CommonUtils.createBlockPath(blockPath);
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFS.java]: Perhaps this should be handled in the worker as well? Though does not have to be in this PR. Then createBlockPath can go into the (not yet merged) BlockWorkerUtils?

File: servers/src/main/java/tachyon/worker/netty/DataServerHandler.java [hunk 12]
```diff
@@ -126,6 +137,48 @@ private void handleBlockRequest(final ChannelHandlerContext ctx, final RPCBlockR
...
+          ctx.writeAndFlush(new RPCBlockWriteResponse(userId, blockId, offset, length, true));
+      future.addListener(ChannelFutureListener.CLOSE);
+    } catch (Exception e) {
+      LOG.error("Error writing remote block : " + e.getMessage(), e);
+      RPCBlockWriteResponse resp =
+          new RPCBlockWriteResponse(userId, blockId, offset, length, false);
+      ChannelFuture future = ctx.writeAndFlush(resp);
+      future.addListener(ChannelFutureListener.CLOSE);
+    }
+  }
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/netty/DataServerHandler.java]: Does the writer need to be closed?

File: clients/unshaded/src/main/java/tachyon/client/TachyonFS.java (hunk 11 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFS.java]: Ok, I added a TODO for this.

File: servers/src/main/java/tachyon/worker/netty/DataServerHandler.java (hunk 12 above)
Comment (gpang) [inline servers/src/main/java/tachyon/worker/netty/DataServerHandler.java]: The writer is being closed now. Thanks!

**Comment by gpang** [comment]
Thanks for the comments @calvinjia . I addressed them and added some more test cases.

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java [hunk 13]
```diff
@@ -16,239 +16,63 @@
...
-    if (mBuffer.position()
-        >= mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB)) {
-      appendCurrentBuffer(mBuffer.array(), 0, mBuffer.position());
-      mBuffer.clear();
-    }
-
-    CommonUtils.putIntByteBuffer(mBuffer, b);
-    mWrittenBytes ++;
-  }
+  public abstract long getRemainingSpaceByte();
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Please javadoc this method since it is public.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+   * @param initialBytes the initial size bytes that will be allocated to the block
+   * @param tachyonConf the TachyonConf instance for this file output stream.
+   * @throws IOException
+   */
+  LocalBlockOutStream(TachyonFile file, WriteType opType, int blockIndex, long initialBytes,
+                      TachyonConf tachyonConf) throws IOException {
+    super(file, opType, tachyonConf);
+
+    if (!opType.isCache()) {
+      throw new IOException("LocalBlockOutStream only support WriteType.CACHE");
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: stating the writeType which triggers this exception in message?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+    // Open the remote writer.
+    mRemoteWriter =
+        mCloser.register(RemoteBlockWriter.Factory.createRemoteBlockWriter(tachyonConf));
+    mRemoteWriter.open(mTachyonFS.getWorkerDataServerAddress(), mBlockId, mTachyonFS.getUserId());
+    mOpen = true;
+  }
+
+  // Write data to the remote block.
+  private synchronized void writeToRemoteBlock(byte[] bytes, int offset, int length)
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: javadoc why we need `synchronized` here

File: integration-tests/src/test/java/tachyon/client/BlockOutStreamIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+import tachyon.Constants;
+import tachyon.TachyonURI;
+import tachyon.conf.TachyonConf;
+import tachyon.master.LocalTachyonCluster;
+
+/**
+ * Integration tests for <code>tachyon.client.BlockOutStream</code>.
+ */
+public class BlockOutStreamIntegrationTest {
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/BlockOutStreamIntegrationTest.java]: If `LocalTachyonCluster` is unnecessary here (by creating `TachyonConf` directly), we may move this test into unittest.

File: servers/src/main/java/tachyon/worker/block/BlockDataManager.java [hunk 17]
```diff
@@ -243,7 +243,22 @@ public String createBlock(long userId, long blockId, int tierAlias, long initial
   public BlockWriter createBlockRemote(long userId, long blockId, int tierAlias, long initialBytes)
       throws FileDoesNotExistException, IOException {
     BlockStoreLocation loc = BlockStoreLocation.anyDirInTier(tierAlias);
-    mBlockStore.createBlockMeta(userId, blockId, loc, initialBytes);
+    TempBlockMeta createdBlock = mBlockStore.createBlockMeta(userId, blockId, loc, initialBytes);
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/BlockDataManager.java]: Currently it is a little bit confusing because both `createBlockRemote` and `getTempBlockWriterRemote` returns `BlockWriter`. I think we may change this method `createBlockRemote` to only create a block---without returning a `BlockWriter`. And we always use `getTempBlockWriterRemote` to return the `BlockWriter`

File: servers/src/main/java/tachyon/worker/netty/DataServerHandler.java [hunk 18]
```diff
@@ -126,6 +137,52 @@ private void handleBlockRequest(final ChannelHandlerContext ctx, final RPCBlockR
...
+
+    BlockWriter writer = null;
+    try {
+      req.validate();
+      ByteBuffer buffer = data.getReadOnlyByteBuffer();
+      if (buffer.remaining() <= 0) {
+        throw new IOException("Empty buffer to write.");
+      }
+
+      if (offset == 0) {
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/netty/DataServerHandler.java]: If we change the semantics of `createBlockRemote` by not returning `BlockWriter`, then we could probably simplify the code here a bit.

**Comment by apc999** [comment]
Thanks @gpang for the update. I left a bunch of comments---mostly readability issues.

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 13 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: done.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 14 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Added the opType in the message.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java (hunk 15 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Added comment.

File: integration-tests/src/test/java/tachyon/client/BlockOutStreamIntegrationTest.java (hunk 16 above)
Comment (gpang) [inline integration-tests/src/test/java/tachyon/client/BlockOutStreamIntegrationTest.java]: It turns out the local cluster is necessary, since in order to create a file, it needs to contact the master.

File: servers/src/main/java/tachyon/worker/block/BlockDataManager.java (hunk 17 above)
Comment (gpang) [inline servers/src/main/java/tachyon/worker/block/BlockDataManager.java]: Yes, this is better. I changed it.

File: servers/src/main/java/tachyon/worker/netty/DataServerHandler.java (hunk 18 above)
Comment (gpang) [inline servers/src/main/java/tachyon/worker/netty/DataServerHandler.java]: It is only slightly less complicated, since space allocation is different in each case.

**Comment by gpang** [comment]
Thanks @apc999 for the comments! I addressed them and updated the PR.

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java [hunk 19]
```diff
@@ -16,239 +16,66 @@
...
-      if (mBuffer.position() > 0) {
-        appendCurrentBuffer(mBuffer.array(), 0, mBuffer.position());
-      }
-      mCloser.close();
-      mTachyonFS.cacheBlock(mBlockId);
-      mClosed = true;
+  public static BlockOutStream get(TachyonFile tachyonFile, WriteType opType, int blockIndex,
+      long initialBytes, TachyonConf tachyonConf) throws IOException {
+
+    if (tachyonFile.mTachyonFS.hasLocalWorker()
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Is `hasLocalWorker` a strong enough condition? Its possible the local worker cannot allocate space for the request?

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java [hunk 20]
```diff
@@ -16,239 +16,66 @@
...
+        && tachyonConf.getBoolean(Constants.USER_ENABLE_LOCAL_WRITE,
+            Constants.DEFAULT_USER_ENABLE_LOCAL_WRITE)) {
+      LOG.info("Writing with local stream.");
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Could we add more detail to the two `LOG.info` statements (L65/69)?

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 13 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Maybe its better to name it `getRemainingSpaceBytes`?

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    mBlockIndex = blockIndex;
+    mBlockCapacityByte = mFile.getBlockSizeByte();
+    mBlockId = mFile.getBlockId(mBlockIndex);
+    mBlockOffset = mBlockCapacityByte * blockIndex;
+
+    if (!mTachyonFS.hasLocalWorker()) {
+      throw new IOException("The machine does not have any local worker.");
+    }
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Is this check still necessary since we do everything through the `get` method?

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+
+    // TODO: Use the new LocalFileBlockWriter api for writing local files.
+    mLocalFilePath = mTachyonFS.getLocalBlockTemporaryPath(mBlockId, initialBytes);
+    mLocalFile = mCloser.register(new RandomAccessFile(mLocalFilePath, "rw"));
+    mLocalFileChannel = mCloser.register(mLocalFile.getChannel());
+    // change the permission of the temporary file in order that the worker can move it.
+    CommonUtils.changeLocalFileToFullPermission(mLocalFilePath);
+    // use the sticky bit, only the client and the worker can write to the block
+    CommonUtils.setLocalFileStickyBit(mLocalFilePath);
+    LOG.info(mLocalFilePath + " was created!");
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Is there any more info we could add to this log statement?

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+    mAvailableBytes += initialBytes;
+
+    long allocateBytes = mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB) + 4L;
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Do we need the `+ 4L`? Also maybe make the `userFileBufferBytes` an instance variable since we use it multiple times (and the conf is presumably constant for now)?

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+      }
+      mCloser.close();
+      mTachyonFS.cacheBlock(mBlockId);
+      mClosed = true;
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    // Since this only writes to memory, this flush is not outside visible.
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Does it make sense to just append the current buffer if it exists?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java [hunk 25]
```diff
@@ -0,0 +1,224 @@
...
+public class RemoteBlockOutStream extends BlockOutStream {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final int mBlockIndex;
+  private final long mBlockId;
+  private final long mBlockCapacityBytes;
+  private final Closer mCloser;
+  private final RemoteBlockWriter mRemoteWriter;
+
+  // The local write buffer to accumulate client writes.
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Minor style comment, perhaps use `/** description */`?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java [hunk 26]
```diff
@@ -0,0 +1,224 @@
...
+  private final ByteBuffer mBuffer;
+  // The size of the write buffer in bytes.
+  private final long mBufferBytes;
+  // Total number of bytes written to block, but not necessarily flushed to the remote server.
+  private long mWrittenBytes = 0;
+  // True if stream is open.
+  private boolean mOpen = false;
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Can we change this to `mClosed` and reverse the logic to be consistent with the rest of the codebase?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+
+  /**
+   * @param file the file the block belongs to
+   * @param opType the OutStream's write type
+   * @param blockIndex the index of the block in the file
+   * @param initialBytes the initial size bytes that will be allocated to the block
+   * @param tachyonConf the TachyonConf instance for this file output stream.
+   * @throws IOException
+   */
+  RemoteBlockOutStream(TachyonFile file, WriteType opType, int blockIndex, long initialBytes,
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Does `initialBytes` ever get used?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    mBlockIndex = blockIndex;
+    mBlockCapacityBytes = mFile.getBlockSizeByte();
+    mBlockId = mFile.getBlockId(mBlockIndex);
+    mCloser = Closer.create();
+
+    // Create a local buffer.
+    mBufferBytes = mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB);
+    long allocateBytes = mBufferBytes;
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Why do we need this variable?

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java [hunk 29]
```diff
@@ -0,0 +1,224 @@
...
+    writeToRemoteBlock(mBuffer.array(), 0, mBuffer.position());
+    mBuffer.clear();
+  }
+
+  @Override
+  public void cancel() throws IOException {
+    if (mOpen) {
+      mCloser.close();
+      mOpen = false;
+      if (mWrittenBytes > 0) {
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: What's the reasoning behind checking `mWrittenBytes` since this does not necessarily mean anything was written remotely?

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 13 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Renamed.

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 19 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: I think if the local worker cannot allocate space, it will fail later on. This condition is the similar condition as before (just threw an exception if there was no local worker).

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 20 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Added more info to the messages.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 21 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: The check is not necessary. I added more comments stating that the get() should be used.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 22 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Added some more details.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 23 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Most (if not all) of this file was just renamed from the original BlockOutStream, so these are artifacts of the existing file. I never knew why there was that extra 4, but I removed it.

I also added an instance variable for the buffer size.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 24 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: Done.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java (hunk 25 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Done.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java (hunk 26 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Done.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java (hunk 27 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: No, so I added a comment about that.

Essentially, since the remote write api is really simple right now, the server side handles the allocation to match the size of the first write. If we ever wanted to add a richer api or add state to the data server, we could use this initialBytes parameter.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java (hunk 28 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: We don't. I removed it.

File: clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java (hunk 29 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/RemoteBlockOutStream.java]: Thanks for catching this! I added a variable keeping track of the flushed bytes.

**Comment by gpang** [comment]
I updated the PR addressing your comments, @calvinjia . Thanks!

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 19 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Oh I meant since we have remote write, the write can possibly still be served, but this is a feature that can be added later.

**Comment by calvinjia** [comment]
Jenkins, test this please.

**Comment by gpang** [comment]
Jenkins, test this please.

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 21 above)
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: But it might be good to have a check here.
how about having `Preconditions.checkState(mTachyonFS.hasLocalWorker())` here?

**Comment by apc999** [comment]
+1 LGTM. Thanks for completing this important feature!

File: clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java (hunk 21 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/LocalBlockOutStream.java]: I added it in case it is constructed directly.

File: clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java (hunk 19 above)
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java]: Yeah, that could be a useful feature, but I don't think it can be done simply right now.

In the future, maybe we can make the client re-connect to a different worker if the local worker does not have space. This may have to touch higher levels of the code, like the TFS, and/or master.

**Comment by gpang** [comment]
Thanks for all the reviews!

**Comment by calvinjia** [comment]
+1 LGTM

**Comment by haoyuan** [comment]
Awesome!
